mistralrs-core/src/engine/mod.rs

```diff
@@ -116,331 +116,340 @@ impl Engine {
+                                            completion_lengths,
+                                            ms_from_last_run * 1000.,
+                                        );
+                                    }
+                                }
 
-                        if is_prompt {
-                            for mut seq in guards {
-                                let now = SystemTime::now()
-                                    .duration_since(UNIX_EPOCH)
-                                    .expect("Time travel has occurred!")
-                                    .as_millis();
+                                let throughput_end = Instant::now();
                                 #[allow(clippy::cast_precision_loss)]
-                                let prompt_tok_per_sec =
-                                    seq.len() as f32 / (now - seq.timestamp()) as f32;
-                                seq.prompt_tok_per_sec = prompt_tok_per_sec * 1000.;
-                                seq.prompt_timestamp = Some(now);
+                                if self.throughput_logging_enabled {
+                                    let n_toks = if is_prompt {
+                                        guards.iter().map(|seq| seq.get_toks().len()).sum::<usize>()
+                                    } else {
+                                        guards.len()
+                                    };
+                                    let ts = n_toks as f64
+                                        / throughput_end
+                                            .duration_since(throughput_start)
+                                            .as_secs_f64();
+                                    info!("Throughput (scheduler V2): {ts} T/s");
+                                }
+
+                                if is_prompt {
+                                    for mut seq in guards {
+                                        let now = SystemTime::now()
+                                            .duration_since(UNIX_EPOCH)
+                                            .expect("Time travel has occurred!")
+                                            .as_millis();
+                                        #[allow(clippy::cast_precision_loss)]
+                                        let prompt_tok_per_sec =
+                                            seq.len() as f32 / (now - seq.timestamp()) as f32;
+                                        seq.prompt_tok_per_sec = prompt_tok_per_sec * 1000.;
+                                        seq.prompt_timestamp = Some(now);
+                                    }
+                                }
                             }
                         }
                     }
-                }
-            }
 
-            self.scheduler.free_finished_sequence_groups();
+                    if self.scheduler.waiting_len() == 0 {
+                        tokio::task::yield_now().await;
```

Comment:
I like the idea of the `tokio::select!` addition! I'm just a bit confused about how this function would implement what we want (that is, to not sit in a loop). If I understand correctly, it yields to tokio's runtime, which means we wait for the other arm of the `tokio::select!` (the request recieve arm) to match?

Perhaps you could add a comment here explaining what the logic/flow is?

Reply:
You understand the flow correctly. The reason we no longer sit in the loop anymore is mostly because of the `yield_now`, it basically marks the task as `Pending` for one iteration of Tokio's runtime. On the next iteration, `yield_now` will marked `Ready` and the loop can continue. Yielding doesn't take much time, but is enough to lower CPU usage dramatically.

Reply:
@EricLBuehler Thoughts on this? Happy to close this and rethink if this change doesn't make sense to you.